File: packages/python-sdk/prism/_budget.py
import os
import time
import warnings
from datetime import datetime
from typing import Dict, Optional, Tuple

import httpx


class BudgetExceededError(Exception):
    """Raised when a hard budget cap is enforced and spend exceeds limit."""


# Module-level cache: key_prefix:month → (result_dict, monotonic_timestamp)
_CACHE: Dict[str, Tuple[dict, float]] = {}
_CACHE_TTL = 30.0  # seconds


def _cache_get(key: str) -> Optional[dict]:
    entry = _CACHE.get(key)
    if entry and time.monotonic() - entry[1] < _CACHE_TTL:
        return entry[0]
    _CACHE.pop(key, None)
    return None


def _cache_set(key: str, data: dict) -> None:
    _CACHE[key] = (data, time.monotonic())


class BudgetChecker:
    """
    Checks the current month's spend against the configured budget by calling
    /api/budget/check. The server resolves the project and org from the API key.
    Results cached for 30s to avoid per-call overhead.
    """

    def __init__(self, prism_key: str) -> None:
        self._prism_key = prism_key
        self._base_url  = os.environ.get("PRISM_APP_URL", "https://useprism.dev").rstrip("/")
        self._http      = httpx.Client(timeout=2.0)

    def check_or_raise(self) -> None:
        cache_key = f"{self._prism_key[:12]}:{datetime.now().strftime('%Y-%m')}"
        result    = _cache_get(cache_key)

        if result is None:
            result = self._fetch()
            if result is not None:
                _cache_set(cache_key, result)

        if result is not None and not result.get("allowed", True):
            spend = result.get("spend", 0)
            limit = result.get("limit", 0)
            raise BudgetExceededError(
                f"Monthly budget exceeded: ${spend:.4f} spent of ${limit:.4f} limit. "
                "Set a higher budget in the Prism dashboard or disable enforce_hard_cap."
            )

    def _fetch(self) -> Optional[dict]:
        try:
            resp = self._http.get(
                f"{self._base_url}/api/budget/check",
                headers={"Authorization": f"Bearer {self._prism_key}"},
            )
            if resp.status_code == 401:
                warnings.warn("Prism: invalid API key for budget check", stacklevel=3)
                return None
            if not resp.is_success:
                return None
            return resp.json()
        except Exception as exc:
            warnings.warn(f"Prism: budget check failed ({exc})", stacklevel=3)
            return None

    def close(self) -> None:
        self._http.close()

File: packages/typescript-sdk/src/budget.ts
export class BudgetExceededError extends Error {
  readonly spend: number;
  readonly limit: number;

  constructor(spend: number, limit: number) {
    super(
      `Monthly budget exceeded: $${spend.toFixed(4)} spent of $${limit.toFixed(4)} limit. ` +
      "Set a higher budget in the Prism dashboard or disable enforce_hard_cap.",
    );
    this.name  = "BudgetExceededError";
    this.spend = spend;
    this.limit = limit;
  }
}

interface BudgetResult {
  allowed: boolean;
  spend:   number;
  limit:   number | null;
  enforce: boolean;
  pct:     number;   // spend as % of limit (0 when no limit)
}

/** Status returned by checkStatus() */
export type BudgetStatusResult = "ok" | "soft_cap_hit" | "hard_cap_exceeded";

const cache = new Map<string, { result: BudgetResult; expiresAt: number }>();
const CACHE_TTL_MS = 30_000;

function cacheGet(key: string): BudgetResult | null {
  const entry = cache.get(key);
  if (!entry) return null;
  if (Date.now() > entry.expiresAt) { cache.delete(key); return null; }
  return entry.result;
}

function cacheSet(key: string, result: BudgetResult): void {
  cache.set(key, { result, expiresAt: Date.now() + CACHE_TTL_MS });
}

export class BudgetChecker {
  private readonly key:     string;
  private readonly baseUrl: string;

  constructor(key: string) {
    this.key     = key;
    this.baseUrl = (
      process.env["PRISM_APP_URL"] ??
      process.env["NEXT_PUBLIC_APP_URL"] ??
      "https://useprism.dev"
    ).replace(/\/$/, "");
  }

  /**
   * Returns the current budget status without throwing.
   * Used internally for the soft-cap model downgrade flow.
   * @param softCapPct - percentage threshold for soft cap (default 80)
   */
  async checkStatus(softCapPct = 80): Promise<BudgetStatusResult> {
    const month    = new Date().toISOString().slice(0, 7);
    const cacheKey = `${this.key.slice(0, 12)}:${month}`;

    let result = cacheGet(cacheKey);
    if (result === null) {
      result = await this._fetch();
      if (result !== null) cacheSet(cacheKey, result);
    }

    if (result === null) return "ok"; // API down — fail open

    if (!result.allowed) return "hard_cap_exceeded";
    if (result.limit != null && result.limit > 0 && result.pct >= softCapPct) {
      return "soft_cap_hit";
    }
    return "ok";
  }

  async checkOrThrow(): Promise<void> {
    const status = await this.checkStatus(101); // only hard cap
    if (status === "hard_cap_exceeded") {
      const month    = new Date().toISOString().slice(0, 7);
      const cacheKey = `${this.key.slice(0, 12)}:${month}`;
      const result   = cacheGet(cacheKey);
      throw new BudgetExceededError(result?.spend ?? 0, result?.limit ?? 0);
    }
  }

  private async _fetch(): Promise<BudgetResult | null> {
    try {
      const res = await fetch(`${this.baseUrl}/api/budget/check`, {
        headers: { Authorization: `Bearer ${this.key}` },
      });
      if (res.status === 401) {
        console.warn("[prism] Invalid API key for budget check");
        return null;
      }
      if (!res.ok) return null;
      return (await res.json()) as BudgetResult;
    } catch {
      return null;
    }
  }
}
